File: backend/app/api/v1/projects.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_session
from app.schemas.project import ProjectCreate, ProjectResponse, ProjectTaskResponse, ProjectUpdate
from app.services import project_service

router = APIRouter(prefix="/projects", tags=["projects"])


@router.get("", response_model=list[ProjectResponse])
async def list_projects(session: AsyncSession = Depends(get_session)):
    projects = await project_service.list_projects(session)
    result = []
    for p in projects:
        count = await project_service.get_task_count(p.id, session)
        result.append(ProjectResponse(
            id=p.id,
            title=p.title,
            description=p.description,
            notes=p.notes,
            color=p.color,
            created_at=p.created_at,
            task_count=count,
        ))
    return result


@router.post("", response_model=ProjectResponse, status_code=201)
async def create_project(body: ProjectCreate, session: AsyncSession = Depends(get_session)):
    proj = await project_service.create_project(body, session)
    return ProjectResponse(
        id=proj.id,
        title=proj.title,
        description=proj.description,
        notes=proj.notes,
        color=proj.color,
        created_at=proj.created_at,
        task_count=0,
    )


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(project_id: str, session: AsyncSession = Depends(get_session)):
    proj = await project_service.get_project(project_id, session)
    if not proj:
        raise HTTPException(status_code=404, detail="Project not found")
    count = await project_service.get_task_count(project_id, session)
    return ProjectResponse(
        id=proj.id,
        title=proj.title,
        description=proj.description,
        notes=proj.notes,
        color=proj.color,
        created_at=proj.created_at,
        task_count=count,
    )


@router.put("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: str, body: ProjectUpdate, session: AsyncSession = Depends(get_session)
):
    proj = await project_service.update_project(project_id, body, session)
    if not proj:
        raise HTTPException(status_code=404, detail="Project not found")
    count = await project_service.get_task_count(project_id, session)
    return ProjectResponse(
        id=proj.id,
        title=proj.title,
        description=proj.description,
        notes=proj.notes,
        color=proj.color,
        created_at=proj.created_at,
        task_count=count,
    )


@router.delete("/{project_id}", status_code=204)
async def delete_project(project_id: str, session: AsyncSession = Depends(get_session)):
    deleted = await project_service.delete_project(project_id, session)
    if not deleted:
        raise HTTPException(status_code=404, detail="Project not found")


@router.get("/{project_id}/tasks", response_model=list[ProjectTaskResponse])
async def get_project_tasks(project_id: str, session: AsyncSession = Depends(get_session)):
    tasks = await project_service.get_project_tasks(project_id, session)
    return [
        ProjectTaskResponse(
            id=t.id,
            title=t.title,
            status=t.status.value,
            priority=t.priority.value,
            due_date=t.due_date.isoformat() if t.due_date else None,
        )
        for t in tasks
    ]


@router.post("/{project_id}/tasks/{task_id}", status_code=204)
async def assign_task(
    project_id: str, task_id: str, session: AsyncSession = Depends(get_session)
):
    task = await project_service.assign_task_to_project(task_id, project_id, session)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")


@router.delete("/{project_id}/tasks/{task_id}", status_code=204)
async def remove_task(
    project_id: str, task_id: str, session: AsyncSession = Depends(get_session)
):
    task = await project_service.assign_task_to_project(task_id, None, session)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

File: frontend/src/api/projects.ts
import client from './client'

export interface Project {
  id: string
  title: string
  description: string | null
  notes: string | null
  color: string
  created_at: string
  task_count: number
}

export interface ProjectTask {
  id: string
  title: string
  status: string
  priority: string
  due_date: string | null
}

export async function getProjects(): Promise<Project[]> {
  const { data } = await client.get<Project[]>('/projects')
  return data
}

export async function createProject(payload: {
  title: string
  description?: string
  color?: string
}): Promise<Project> {
  const { data } = await client.post<Project>('/projects', payload)
  return data
}

export async function getProject(id: string): Promise<Project> {
  const { data } = await client.get<Project>(`/projects/${id}`)
  return data
}

export async function updateProject(
  id: string,
  payload: Partial<{ title: string; description: string; notes: string; color: string }>
): Promise<Project> {
  const { data } = await client.put<Project>(`/projects/${id}`, payload)
  return data
}

export async function deleteProject(id: string): Promise<void> {
  await client.delete(`/projects/${id}`)
}

export async function getProjectTasks(id: string): Promise<ProjectTask[]> {
  const { data } = await client.get<ProjectTask[]>(`/projects/${id}/tasks`)
  return data
}

export async function assignTaskToProject(projectId: string, taskId: string): Promise<void> {
  await client.post(`/projects/${projectId}/tasks/${taskId}`)
}

export async function removeTaskFromProject(projectId: string, taskId: string): Promise<void> {
  await client.delete(`/projects/${projectId}/tasks/${taskId}`)
}
